### ISSUE TITLE
two issue on run YCSB

### ISSUE THREAD

**Issue by Hadi14** [brianfrankcooper/YCSB#606]
Hi 
excusme , on my Loads and runs i hit to two issue
1- i load about 1 Million operation on WorkloadA,F and Redid databse with threads 1,2,4 but load action is waiting  and not complete Load (waiting on ```starting test```), what is problem in your opinion? (my system has only 2G of Ram , it should be it's reason correct?)

2-on other , YCSB Load properly  but in run i got like this timeout error:
```
redis.clients.jedis.exceptions.JedisConnectionException: java.net.SocketTimeoutException: Read timed out
	at redis.clients.jedis.Protocol.process(Protocol.java:74)
	at redis.clients.jedis.Protocol.read(Protocol.java:122)
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:178)
	at redis.clients.jedis.Jedis.zadd(Jedis.java:1448)
	at com.yahoo.ycsb.db.RedisClient.insert(RedisClient.java:114)
	at com.yahoo.ycsb.DBWrapper.insert(DBWrapper.java:158)
	at com.yahoo.ycsb.workloads.CoreWorkload.doInsert(CoreWorkload.java:549)
	at com.yahoo.ycsb.ClientThread.run(Client.java:346)
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:152)
	at java.net.SocketInputStream.read(SocketInputStream.java:122)
	at java.net.SocketInputStream.read(SocketInputStream.java:108)
	at redis.clients.util.RedisInputStream.fill(RedisInputStream.java:110)
	at redis.clients.util.RedisInputStream.readByte(RedisInputStream.java:46)
	at redis.clients.jedis.Protocol.process(Protocol.java:59)
	... 7 more
```
thankyou

**Comment by kruthar** [comment]
@Hadi14 - again, can't help you if we don't have any information on what you are trying to do. Will need to see your YCSB command, pertinent workload files and output printouts.

Also, please focus each thread on one single issue. You have posted two here, but not given enough information or context to solve either.

**Comment by Hadi14** [comment]
@kruthar ,for question  1 my command line is :
```bin/ycsb load redis -P workloads/workloadf -p recordcount=1024000 -p "redis.host=127.0.0.1" -p "redis.port=6379" > ~/result/workload_res.txt```

**Comment by kruthar** [comment]
Can you please attach the whole output....

**Comment by Hadi14** [comment]
for 1 it stay on ```starting test``` and no output msg.

**Comment by busbey** [comment]
How long have you waited? That workload does a significant amount of set up work. It's supposed to print a warning that it can take several minutes.

**Comment by Hadi14** [comment]
i run for  several times and i dont know how much exactly but in compare for other run and it was for long time.

**Comment by kruthar** [comment]
@Hadi14 - couple things:
* For now let's stick with question 1
* Can you rerun your command without redirecting to a file, this is part of the reason why you are seeing less output. Please post all output you get from that command here.
* Can you also try running with with the ```-s``` flag to see if you are getting any progress at all?

You are trying to insert over 1,000,000 records. Even if every insert only too 1ms, it would still take you +16hours to insert that many records on a single thread. So, if you are routing all of the output to a file (if there is any relevant output to begin with) then this would look like it is hanging but it is actually just working very hard.

Also - can you check redis to see if records are being added?

**Comment by Hadi14** [comment]
@kruthar ok , but i think question 2 is also important for my benchmark correct?
yes  i'll run without redirect to file


but with 1000000 record if every take 1ms , it take 16 minute correct?
excusme how to check how many is added?

**Comment by kruthar** [comment]
Yes, 16 minutes, sorry bad math on my part.

If you run with the ```-s`` flag you can view status as the workload is running and see if any records are being added.

You will have to research how to query Redis directly and check number of records.

**Comment by Hadi14** [comment]
one note , when i use 
command line is:
```Command line: -db com.yahoo.ycsb.db.CassandraCQLClient -p hosts=localhost -p recordcount=1024000 -P workloads/workloada -load```

and content error:
<pre><code>Datacenter: datacenter1; Host: localhost/127.0.0.1; Rack: rack1
[OVERALL], RunTime(ms), 260219.0
[OVERALL], Throughput(ops/sec), 1814.633059077162
[CLEANUP], Operations, 1.0
[CLEANUP], AverageLatency(us), 33520.0
[CLEANUP], MinLatency(us), 33504.0
[CLEANUP], MaxLatency(us), 33535.0
[CLEANUP], 95thPercentileLatency(us), 33535.0
[CLEANUP], 99thPercentileLatency(us), 33535.0
[INSERT], Operations, 472203.0
[INSERT], AverageLatency(us), 543.6595934375682
[INSERT], MinLatency(us), 327.0
[INSERT], MaxLatency(us), 2018303.0
[INSERT], 95thPercentileLatency(us), 698.0
[INSERT], 99thPercentileLatency(us), 954.0
[INSERT], Return=OK, 472202
[INSERT], Return=ERROR, 1</code></pre>

**Comment by kruthar** [comment]
@Hadi14 - What you posted doesn't give any information on the error. When you redirect to a file you have to send that information as well, otherwise we can't tell what is going on.

**Comment by busbey** [comment]
The error response just means that the driver returned a failure for one if the operations. If there isn't more detail in the redirected output, you should check the server logs.

**Comment by Hadi14** [comment]
when i run without redirect file it work as well but when i use redirected file it got this error.  why?

above error is all of file content.

**Comment by busbey** [comment]
you'll need to look at the server logs and/or if possible adjust the logging level of the particular client-side driver. doing either of these will require datastore specific instructions; you should consult the docs for the relevant datastore to see where the server logs are and how to change logging levels on the client.

**Comment by Hadi14** [comment]
i dont know why this happen , do you know where are system logs on ubuntu?

**Comment by busbey** [comment]
The location of the system logs depends on how you install Cassandra and how it is configured to do logging.

**Comment by Hadi14** [comment]
excusme , what is reason of event on question2?

**Comment by kruthar** [comment]
>redis.clients.jedis.exceptions.JedisConnectionException: java.net.SocketTimeoutException: Read timed out

This error is likely a connection timeout with Redis when doing an operation. There are a lot of potential reasons for this happening, some are:
- Redis is having an issue and can't respond, or takes to long to respond
- Redis is working fine, but just overloaded and so can't respond quick enough

Do you have this issue when running benchmarks with less records?

**Comment by Hadi14** [comment]
@kruthar i have 2 machine , one 2GRam and Ci3 processor 2.53GHZ and WorkloadF with 1Thread and 512000 record (on redis) ,  other one 32GRam and 16 core 2.29GHZ both WorkloadA and WorkloadF on cassandra got this error.

**Comment by kruthar** [comment]
The Redis client uses Jedis. It looks like the default connection timeout is 2 seconds: https://github.com/xetorthio/jedis/wiki/FAQ. Unfortunately, it does not look like that client does not expose the timeout to be configured, feel free to update the client and try with a long timeout.

On the machine that has Redis installed. Can you run the ```load``` command from YCSB to get all of your records into Redis, then use the redis-cli to test some gets and sets to see how long it takes you?

**Comment by kruthar** [comment]
@Hadi14 - I'm sorry I am not sure what your issue is. I tried your exact command on a VM with 2GB of memory and 4 cpus and had no issues.

Is YCSB on the same machine as Redis when running on your 2GB machine?

**Comment by Hadi14** [comment]
@kruthar yes you say right default connection timeout is 2 seconds but also in this link
http://stackoverflow.com/questions/14993644/configure-jedis-timeout
is said that "redis will close idle client connections after 60 seconds"

>Can you run the load command from YCSB to get all of your records into Redis, then use the redis-cli to test some gets and sets to see how long it takes you?
yes all 512000 record are loaded , this error got on run action.

>Is YCSB on the same machine as Redis when running on your 2GB machine?
yes


- on other machine i got this error on `cassandra` , what is timeout for cassandra?

**Comment by kruthar** [comment]
I meant you should see if ```get``` in the redis-cli is also slow. the redis-cli has nothing to do with YCSB, it is the command line interface (cli) that comes with your Redis installation. If you are not sure what that is, this quickstart article at Redis details how to use the redis-cli a little bit: http://redis.io/topics/quickstart.

Let's stay focused on your issues with Redis for now, it is very confusing trying to solve two different issues with two different clients at the same time, looking back through this thread I can't tell sometimes which issue you are referring to.

Can you please send the *ENTIRE* output *DO NOT* redirect to a file of a ```run``` command that fails with the issues you are seeing.

When you run a benchmark against Redis that is much smaller, say 10,000 records and 10,000 operations, do you have the same issues?

**Comment by Hadi14** [comment]
>this quickstart article at Redis details how to use the redis-cli a little bit: http://redis.io/topics/quickstart.

i installed redis base on above link
>Can you please send the ENTIRE output DO NOT redirect to a file of a run command that fails with the issues you are seeing.

yes ok.

>When you run a benchmark against Redis that is much smaller, say 10,000 records and 10,000 operations, do you have the same issues?

i load and run YCSB with 256000 it work fine on 46s.

**Comment by Hadi14** [comment]
@kruthar 
All error is this:
<pre>`bin/ycsb load redis -P workloads/workloadf -p recordcount=512000 -p "redis.host=127.0.0.1" -p "redis.port=6379"
java -cp /home/hadi/ycsb-0.5.0/redis-binding/conf:/home/hadi/ycsb-0.5.0/conf:/home/hadi/ycsb-0.5.0/lib/core-0.5.0.jar:/home/hadi/ycsb-0.5.0/lib/jackson-mapper-asl-1.9.4.jar:/home/hadi/ycsb-0.5.0/lib/HdrHistogram-2.1.4.jar:/home/hadi/ycsb-0.5.0/lib/jackson-core-asl-1.9.4.jar:/home/hadi/ycsb-0.5.0/redis-binding/lib/jedis-2.0.0.jar:/home/hadi/ycsb-0.5.0/redis-binding/lib/commons-pool-1.5.5.jar:/home/hadi/ycsb-0.5.0/redis-binding/lib/redis-binding-0.5.0.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.RedisClient -P workloads/workloadf -p recordcount=512000 -p redis.host=127.0.0.1 -p redis.port=6379 -load
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.RedisClient -P workloads/workloadf -p recordcount=512000 -p redis.host=127.0.0.1 -p redis.port=6379 -load
Loading workload...
Starting test.
redis.clients.jedis.exceptions.JedisConnectionException: java.net.SocketTimeoutException: Read timed out
	at redis.clients.jedis.Protocol.process(Protocol.java:74)
	at redis.clients.jedis.Protocol.read(Protocol.java:122)
	at redis.clients.jedis.Connection.getStatusCodeReply(Connection.java:152)
	at redis.clients.jedis.Jedis.hmset(Jedis.java:726)
	at com.yahoo.ycsb.db.RedisClient.insert(RedisClient.java:113)
	at com.yahoo.ycsb.DBWrapper.insert(DBWrapper.java:158)
	at com.yahoo.ycsb.workloads.CoreWorkload.doInsert(CoreWorkload.java:549)
	at com.yahoo.ycsb.ClientThread.run(Client.java:346)
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:152)
	at java.net.SocketInputStream.read(SocketInputStream.java:122)
	at java.net.SocketInputStream.read(SocketInputStream.java:108)
	at redis.clients.util.RedisInputStream.fill(RedisInputStream.java:110)
	at redis.clients.util.RedisInputStream.readByte(RedisInputStream.java:46)
	at redis.clients.jedis.Protocol.process(Protocol.java:59)
	... 7 more
redis.clients.jedis.exceptions.JedisConnectionException: java.net.SocketTimeoutException: Read timed out
	at redis.clients.jedis.Protocol.process(Protocol.java:74)
	at redis.clients.jedis.Protocol.read(Protocol.java:122)
	at redis.clients.jedis.Connection.getStatusCodeReply(Connection.java:152)
	at redis.clients.jedis.Jedis.hmset(Jedis.java:726)
	at com.yahoo.ycsb.db.RedisClient.insert(RedisClient.java:113)
	at com.yahoo.ycsb.DBWrapper.insert(DBWrapper.java:158)
	at com.yahoo.ycsb.workloads.CoreWorkload.doInsert(CoreWorkload.java:549)
	at com.yahoo.ycsb.ClientThread.run(Client.java:346)
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:152)
	at java.net.SocketInputStream.read(SocketInputStream.java:122)
	at java.net.SocketInputStream.read(SocketInputStream.java:108)
	at redis.clients.util.RedisInputStream.fill(RedisInputStream.java:110)
	at redis.clients.util.RedisInputStream.readByte(RedisInputStream.java:46)
	at redis.clients.jedis.Protocol.process(Protocol.java:59)
	... 7 more`</pre>

**Comment by busbey** [comment]
That looks like a timeout from the redis cluster not responding in time. When you use the redis CLI, what do response times look like?

**Comment by Hadi14** [comment]
@busbey i load and run YCSB with 256000 it work fine on 46s.

**Comment by busbey** [comment]
If you mean the timeout for redis, the default for the server side is 60s. However, the client library currently used defaults to a 2s timeout. This leaves us with an effective timeout if the minimum, 2s.

If smaller workloads go fine, I suspect you're just hitting a point where your redis instance can't keep up.

Does YCSB report the problem and continue, or is it crashing?

**Comment by Hadi14** [comment]
>However, the client library currently used defaults to a 2s timeout. This leaves us with an effective timeout if the minimum, 2s.

 what is for , this Two seconds, exactly?

>Does YCSB report the problem and continue, or is it crashing?

no it is crashing.

**Comment by busbey** [comment]
@kruthar mentioned above that the jedis client (how YCSB talks to redis) has a timeout of 2s by default for all operations. Right now, that timeout is not configurable because YCSB doesn't expose the option.

Crashing means we need to do a better job in YCSB of interacting with the jedis library.

Any interest in putting together a patch for fixing either of these things?

**Comment by Hadi14** [comment]
>Any interest in putting together a patch for fixing either of these things?

yes but how?

excusme , and other qusetion , what is follow timeout? , i also got error on this and i increase from 2000 to 4000 it worked .
`# How long the coordinator should wait for writes to complete `
`write_request_timeout_in_ms: 2000`

**Comment by busbey** [comment]
> > Any interest in putting together a patch for fixing either of these things?
> yes but how?

Either change would involve editing the redis binding, `src/main/java/com/yahoo/ycsb/db/RedisClient.java`.

For the timeout, you'd have to check on the Jedis library to see how the timeout is adjusted. Then, you'd make a configuration property, pass it to the appropriate part of the Jedis invocation, and document it.

For the exception, you'd have to use proper try/catch blocks in those places where we interact with Jedis library and return an appropriate non-OK status code.

> excusme , and other qusetion , what is follow timeout? , i also got error on this and i increase from 2000 to 4000 it worked for about 1000000 records .
> ```
> # How long the coordinator should wait for writes to complete
> write_request_timeout_in_ms: 2000
> ```
> while i had runtime 476,893ms for 512000 record.

It's some internal detail for Redis. You should check their documentation to find out the specifics of what it means. It sounds like a further indicator that your Redis instance isn't configured such that it can keep up with the load you're asking for in YCSB.

**Comment by Hadi14** [comment]
># How long the coordinator should wait for writes to complete
write_request_timeout_in_ms: 2000

this is on cassandra config file (cassandra.yaml) ,  i increase from 2000 to 4000 it worked for about 1000000 records.

**Comment by busbey** [comment]
> > #How long the coordinator should wait for writes to complete
> > write_request_timeout_in_ms: 2000
> this is on cassandra config file (cassandra.yaml) , i increase from 2000 to 4000 it worked for about 1000000 records.

Please keep a given issue thread focused on one topic. AFAICT, this is the "issues with redis" thread.

My previous statement still largely holds: that value is an internal detail for Cassandra. You should check their documentation to find out specifics. It also sounds like the problem with your Cassandra instance is proper configuration to keep up with the load you're asking for in YCSB.

> i,m sorry other problem , we do benchmark on two machine they have Ascending throughput figure on different records , but both have drop on 128000 record, what do you think on problem regarding on both of machine?

It depends on the datastore you're testing against. Many distributed datastores have points where they'll experience throughput drops due to internal accounting. For example, in a system with implicit sharding such as HBase you can hit a point where one shard has sufficient activity that it must split into two shards. During the brief window where this splitting occurs, throughput will suffer.

If the backing datastore needs to coordinate changes across multiple machines for whatever reason, then passing the threshold form one machine to many might permanently reduce throughput-per-machine, which in turn might reduce overall throughput for a substantial amount of time.

**Comment by kruthar** [comment]
@Hadi14 - can you explain what 'ascending flow' means?

Also, I want to clear up that when you say you are benchmarking against two machines you mean you are working against Redis on your smaller machine, then performing a separate benchmark against Redis on your larger machine. You do not have Redis configured in a cluster across two machines, correct?

How many times did you run your benchmarking test with 64000, 128000, and 256000 records? If you only did one test at each interval, then getting an unexpected result for one test may not actually be representative of what is actually happening. You should perform each test several times to get a clear feel for your results and eliminate outliers.

Also, between your intervals (ex. changing from 64000 to 128000 records) are you clearing out your Redis database and rerunning your ```load``` command with the proper number of records? If you are not, then you could be contaminating your results by having uncertain number of records in your database. Or doing an update when what you are really expecting to do is a fresh insert.

**Comment by Hadi14** [comment]
>can you explain what 'ascending flow' means?

my mean is flow on figure is ascending.
>You do not have Redis configured in a cluster across two machines, correct?

YES
>How many times did you run your benchmarking test with 64000, 128000, and 256000 records?

on redis and first  machin: 9,117 on 64000 and 15,966 on 128000 and 4,713 on 256000 and	 46,546 on 512000 record.

i want know actually what is reason of this happen?

**Comment by kruthar** [comment]
>my mean is flow on figure is ascending.

I understand that the word ascending means that something is rising, but the term ascending flow is confusing. What do you mean by flow, do you mean the average latency through a single test, across tests, do you even mean latency? or something else?

>on redis and first machin: 9,117 on 64000 and 15,966 on 128000 and 4,713 on 256000 and 46,546 on 512000 record.

This actually doesn't answer the question I was asking, I was asking how many times have you run each of those tests? If you have run them multiple times each do you have consistent results?

>i want know actually what is reason of this happen?

I understand you want to know why this is happening, I'm trying to help you figure that out. But to figure that out I have to understand a few things:

* That the results you are reporting here are valid and consistent
* That Redis is installed and configured properly
* If there is a problem, if it is in the YCSB code.

So, far it has been very hard to understand all of this.

**Comment by risdenk** [comment]
@Hadi14 Any update here? were you able to resolve your issue?

**Comment by risdenk** [comment]
Closing as stale.

**Comment by AhmadZia1991** [comment]
please let me know where to run this command ( bin/ycsb load redis -P workloads/workloadf -p recordcount=1024000 -p "redis.host=127.0.0.1" -p "redis.port=6379" > ~/result/workload_res.txt )

**Comment by AhmadZia1991** [comment]
[INFO] Building tar: C:\Users\Ah.Zia Atal\YCSB\redis\target\ycsb-redis-binding-0.16.0-SNAPSHOT.tar.gz
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] YCSB Root 0.16.0-SNAPSHOT .......................... SUCCESS [04:24 min]
[INFO] Core YCSB .......................................... SUCCESS [01:32 min]
[INFO] Per Datastore Binding descriptor ................... SUCCESS [  0.391 s]
[INFO] YCSB Datastore Binding Parent ...................... SUCCESS [05:50 min]
[INFO] Redis DB Binding 0.16.0-SNAPSHOT ................... SUCCESS [ 44.087 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 12:32 min
[INFO] Finished at: 2018-08-27T13:48:48+05:30
[INFO] ------------------------------------------------------------------------



but now i dont know where to run that command
 bin/ycsb load redis -P workloads/workloadf -p recordcount=1024000 -p "redis.host=127.0.0.1" -p "redis.port=6379" > ~/result/workload_res.txt )

**Comment by busbey** [comment]
@AhmadZia1991 please open a new issue if you're having trouble. closed issues don't get seen by many.
